### ISSUE TITLE
Parsing issues since 4.4.0

### ISSUE THREAD

**Issue by neongrau** [presidentbeef/brakeman#1306]
### Background

Brakeman version: 4.4.0
Rails version: 5.1.5
Ruby version:  2.4.4p296 (2018-03-28 revision 63013) [x64-mingw32]

Compared to 4.3.1 which ran without errors and warnings since 4.4.0 after the scan reported no warnings i get 3 errors that make no sense to me.

E.g. an ActiveRecord Model at first line
```
 parse error on value "applications" (tIDENTIFIER)
```
But that line is a comment having the word "applications". When i delete that word the error disappears but brakeman forgets some previous ignores. When re-instating the ignores all is fine.

On 2 Controllers i get warnings about 
```
parse error on value ":" (tCOLON)
```
which are both lines like `render json: { ... }`
changing these to `render :json => { ... }` i still get an error about `kEND` at the end of the file.
Both checks with `ruby -c` and `ruby_parse` on all 3 of these files show no issues.
Yet `brakeman --debug` claims `Could not parse` on those.

**Comment by lumaxis** [comment]
We're seeing very similar issues since 4.4.0. For us, all of them look like this:
```
parse error on value ["do", 18] (kDO_BLOCK)
```

The line it's choking on here looks like this:
```ruby
query do
```
where `query` is a local variable.

**Comment by presidentbeef** [comment]
Thank you for reporting these issues.

It is very hard to figure out parsing problems without example code to reproduce the issue. The parse errors are especially unhelpful because Brakeman will "fall back" on older and older versions of Ruby syntax, so the final error may have nothing to do with the actual problem.

Please try to reduce the files in question down to a minimal example, then anonymize the variables/literals, etc.

@neongrau Are you running `ruby_parse` via RubyParser 3.12.0? Pretty odd to see any difference between `ruby_parse` and Brakeman if they are using the same version, unless you are seeing issues in templates instead of Ruby files.

@lumaxis I guess you mean "where `query` is *also* a local variable" as well as being a method name?  Sounds like you are seeing the same issue as https://github.com/seattlerb/ruby_parser/pull/272

**Comment by lumaxis** [comment]
@presidentbeef That sounds right, yea. I didn't really look closely at it before but here's a simplified version of our code which reproduces the issue in `ruby_parser` `3.12.0`:

```ruby
module StuffConcerns
  module Search
    extend ActiveSupport::Concern

    module ClassMethods
      def search_paginated(query, thing)
        query do
        end
      end
    end
  end
end
```

```
test.rb:7 :: parse error on value ["do", 7] (kDO_BLOCK)> for test.rb
```

**Comment by neongrau** [comment]
@presidentbeef I only have `ruby_parser 3.12.0` and using the cli tool `ruby_parse ./app/controllers/portal_controller.rb` i got no errors.

But using it from a rails console:

    irb(main):005:0> require 'ruby_parser'
    irb(main):006:0> RubyParser.new.parse File.open('app/controllers/portal_controller.rb').read
    Racc::ParseError: (string):18 :: parse error on value ":" (tCOLON)
    irb(main):007:0> RubyParser.for_current_ruby.parse File.open('app/controllers/portal_controller.rb').read
    Racc::ParseError: (string):309 :: parse error on value ["end", 309] (kEND)

While when doing exactly the same on a plain `irb` console i too get no errors.
So there seems to be something from Rails interfering here.

**Comment by neongrau** [comment]
could be related to https://github.com/seattlerb/ruby_parser/pull/272
because the reported errors about `tCOLON` are like this: `render json: { text: text.join('\n') }`
so with the JSON like Hash Syntax ruby_parser doesn't seem to like the key symbol being named like the value containing variable.

But it's getting weirder. ☹️ 

I chose 2 controllers to test with, one that gets reported as having an error and one that doesn't show errors. No difference if i use `RubyParser.for_current_ruby.parse` or `RubyParser.new.parse`

`ControllerGood.rb` -> no error
`ControllerBad.rb` -> tCOLON error

copy content of `ControllerBad.rb` into `ControllerGood.rb`
`ControllerGood.rb` ->  no error

copy content of `ControllerGood.rb` into `ControllerBad.rb`
`ControllerBad.rb` -> tCOLON error

So it's not really the code of the controllers. RubyParser is just randomly complaining about the first colon it finds.

**Comment by presidentbeef** [comment]
Following up here -

The reason why the parse errors are confusing is because Brakeman uses "fallback" behavior in RubyParser. It starts with the newest version of the parser, then works its way back to 1.8 if there are errors. The final error is from the 1.8 parser and is almost always misleading.

I have addressed this in #1320.

@zenspider has been making crazy progress on updated RubyParser, so I expect all the currently reported syntax issues to be addressed in the next release.

**Comment by presidentbeef** [comment]
This is fixed in with ruby_parser 3.13.0 with is included in the Brakeman 4.5.0 release.
